neurons/validator.py

```diff
@@ -651,88 +388,75 @@ def set_weights(self):
-
-        # Update the hotkeys.
-        self.hotkeys = copy.deepcopy(self.metagraph.hotkeys)
+    @property
+    def block(self):
+        return utils.ttl_get_block(self)
 
     def save_state(self):
         """Saves the state of the validator to a file."""
         bt.logging.trace("Saving validator state.")
 
-        if not os.path.exists(self.config.neuron.full_path):
-            os.makedirs(self.config.neuron.full_path)
-
-        # Save the state of the validator to file.
-        self.scorer.save_state(
-            os.path.join(self.config.neuron.full_path, Validator.SCORER_FILENAME)
-        )
+        self.evaluator.save_state()
 
     def load_state(self):
         """Loads the state of the validator from a file."""
         bt.logging.info("Loading validator state.")
 
-        # Load the state of the validator from file.
-        filepath = os.path.join(self.config.neuron.full_path, Validator.SCORER_FILENAME)
-        if not os.path.exists(filepath):
-            bt.logging.warning("No scorer state file found. Starting from scratch.")
-            return
+        self.evaluator.load_state()
 
-        self.scorer.load_state(filepath)
-        bt.logging.success(f"Loaded scorer state from: {filepath}.")
+    def check_registered(self):
+        # --- Check for registration.
+        if self.wallet.hotkey.ss58_address not in self.metagraph.hotkeys:
+            bt.logging.error(
+                f"Wallet: {self.wallet} is not registered on netuid {self.config.netuid}."
+                f" Please register the hotkey using `btcli subnets register` before trying again."
+            )
+            sys.exit(1)
 
-        # Resize the scorer in case the loaded state is old and missing newly added neurons.
-        self.scorer.resize(self.metagraph.n)
 
+def main():
+    """Main constructs the validator with its dependencies."""
 
-# The main function parses the configuration and runs the validator.
-if __name__ == "__main__":
-    with Validator() as validator:
+    config = create_config(NeuronType.VALIDATOR)
+    subtensor = bt.subtensor(config=config)
+    metagraph = subtensor.metagraph(netuid=config.netuid)
+    wallet = bt.wallet(config=config)
+
+    # First, check the wallet is registered.
+    uid = 0
+    if wallet.hotkey.ss58_address in metagraph.hotkeys:
+        uid = metagraph.hotkeys.index(wallet.hotkey.ss58_address)
+        bt.logging.info(
+            f"Running neuron on subnet: {config.netuid} with uid {uid} using network: {subtensor.chain_endpoint}."
+        )
+    else:
+        bt.logging.warning(
+            f"Hotkey {wallet.hotkey.ss58_address} not found in metagraph. Assuming this is a test."
+        )
+
+    # Create the metagraph syncer and perform the initial sync.
+    metagraph_syncer = MetagraphSyncer(
+        subtensor, config={config.netuid: config.neuron.epoch_length * 12}
+    )
+    # Perform an initial sync of all tracked metagraphs.
+    metagraph_syncer.do_initial_sync()
```

Comment:
Should we do the initial_sync after instantiating the miner evaluator so it gets updated too after adding the tracker?

Reply:
I'm not 100% happy with the metagraph_syncer API but couldn't figure out how to make it better. As is, the initial sync here ensures that the metagraphs are available for the validator and evaluator, both of which request the metagraph in their __init__(). If you have suggestions to improve the API, I'm very open.

Reply:
I see. That seems reasonable for now then.